### PR TITLE
Add Index Operations to Translog from Engine

### LINKED ISSUE #2926: [Segment Replication] Merge to main tracking - Refactoring of existing classes to support segrep.
This is a tracking task for the refactoring tasks outlined in our meta merging issue.
- [x] Move RecoveryIndex to top level class and move inner classes from RecoveryIndex
- [x] Move RecoveryState.Timer to top level
- [x] Change GatedAutoCloseable to AutoCloseableRefCounted
- [x] Create base class ReplicationTarget and refactor RecoveryTarget to inherit from it.
- [x] Engine interface changes for addIndexOperationToTranslog and addDeleteOperationToTranslog

### PR DESCRIPTION
**Pull request by Rishikesh1159**
### Description
This PR will allow us to decouple adding Index operations to Lucene and adding Index operations to Translog. This would be useful for segment replication, as we will need to just add Index operation to replicas translog instead of performing full index operation. We will only use this method when segment replication is enabled and when we are on a replica node. The logic for checking if segment replication is enabled and is replica node will be added later in a different PR.

This is a part of the process of merging our feature branch - feature/segment-replication - back into main by re-PRing our changes from the feature branch.
The breakdown of the plan to merge to main is detailed here: https://github.com/opensearch-project/OpenSearch/issues/2926
For added context on segment replication - here's the design proposal https://github.com/opensearch-project/OpenSearch/issues/2229
 
### Issues Resolved
#2926 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kartg** [comment]
Looks good to me! 🚀  My one nitpick/request would be to add javadocs to the new API in `Engine`

I'm stopping short of an approval since I'd like a fresh set of eyes on this 😄

**Comment by Rishikesh1159** [comment]
> Looks good to me! 🚀 My one nitpick/request would be to add javadocs to the new API in `Engine`
> 
> I'm stopping short of an approval since I'd like a fresh set of eyes on this 😄

Sure. I will update the new API with java docs

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -1075,6 +1059,44 @@ public IndexResult index(Index index) throws IOException {
         }
     }
 
+    @Override
+    public Engine.IndexResult addIndexOperationToTranslog(Index index) throws IOException {
+        try (Releasable ignored = versionMap.acquireLock(index.uid().bytes())) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/engine/InternalEngine.java:1064]: I believe you need to acquire release lock here (as all other public methods do):

```
try (ReleasableLock ignored = readLock.acquire()) {
 ....
}
```

Also, it probably makes sense to do basic checks like `ensureOpen();` and `assertIncomingSequenceNumber` before adding anything to translog.

```
ensureOpen();
assert assertIncomingSequenceNumber(index.origin(), index.seqNo());
```

Another concern I have is that write operations like `index` and `delete` do update `lastWriteNanos` but this is not done here, is there a reason?

Comment (nknize) [reply server/src/main/java/org/opensearch/index/engine/InternalEngine.java:1064]: I'd actually prefer to leave this method out until a segrep PR needs it. Nothing in this PR, except tests, call this method so it doesn't make sense to add at this point until we have a change that's actually using it. That will also concretely communicate why the method is needed to those not intimately working segrep; along with how to best structure the method.

**Review by nknize — CHANGES_REQUESTED** [review]
I like the forward thinking approach but lets only introduce methods actually being used at this point and hold on those that are needed in a future PR.

File: server/src/main/java/org/opensearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (reta) [reply server/src/main/java/org/opensearch/index/engine/InternalEngine.java:1064]: :+1: no objections, certainly

**Comment by kartg** [comment]
> I like the forward thinking approach but lets only introduce methods actually being used at this point and hold on those that are needed in a future PR.

The aim of these PRs was to front-load refactoring in existing classes that we've made in the `feature/segment-replication` feature branch. This way,  upcoming PRs on segment replication can focus on functional changes instead of bundling in changes to underlying classes.

If we leave out the `addIndexOperationToTranslog` API in `Engine`, then this PR boils down to simply extracting the `addIndexOperationToTranslog(Index index, IndexResult indexResult)` private method, which doesn't add value in isolation 😅  Thus, exposing an API but not wiring up any callers was our attempt to split the difference.

@nknize any suggestions on how we could balance this better?

**Comment by nknize** [comment]
> any suggestions on how we could balance this better?

We should merge #2904 behind a feature flag (e.g., `opensearch.segment_replication_feature_flag_enabled`) and then work on merging segrep to main so we can do away with the feature branch and continue feature development on main.  Main is meant to be the (unstable) working branch that is constantly moving so I'm not concerned about half baked features so long as they're feature flagged.

**Comment by kotwanikunal** [comment]
@Rishikesh1159 Any updates on these? Are you still blocked?

**Pull request closed without merging**